SSL Client Certificate support

dachaac opened the pull request:
Add support to acquire SSL Client Certificate using socket's sslSession().
`Certificate[] peerCertificates = endpoint.sslSession().getPeerCertificates();`

Support for sslSession() requires latest snapshot of vert.x (https://github.com/eclipse/vert.x/commit/2dcbfac237aed48dd0d4dbeb58372be0ad59c4eb).

In addition to sslSession() one needs to ask for SSL Client Certificates, so added support to MqttServerOptions to specify setClientAuth().

Then for convenience added support for MqttEndpoint#isSsl(), MqttEndpoint#localAddress() and MqttEndpoint#remoteAddress().

With these changes in my test application I was able to identify connecting MQTT client based on its SSL Client Certificate.

dachaac commented:
@ppatierno How does this look for you? Are there something that should be added? Or changed?

ppatierno commented on src/main/java/io/vertx/mqtt/impl/MqttEndpointImpl.java:
```diff
@@ -653,6 +655,32 @@ public void close() {
 
+  public SocketAddress localAddress() {
+    synchronized (this.conn) {
+      return conn.localAddress();
```
any reason why the `checkClosed()` call is missing here ?

dachaac replied:
Hmm... I believe there was similar code without it. Now looking again it seems those codes were protocol message handlers and all commanding codes seem to have it. Adding.

ppatierno commented on src/main/java/io/vertx/mqtt/impl/MqttEndpointImpl.java:
```diff
@@ -653,6 +655,32 @@ public void close() {
+
+  public SocketAddress remoteAddress() {
+    synchronized (this.conn) {
+      return conn.remoteAddress();
```
Ditto as above

dachaac replied:
Adding.

ppatierno commented on src/main/java/io/vertx/mqtt/MqttServerOptions.java:
```diff
@@ -100,6 +101,12 @@ public MqttServerOptions setHost(String host) {
+  public MqttServerOptions setClientAuth(ClientAuth clientAuth) {
+    super.setClientAuth(clientAuth);
+    return this;
+  }
```
Now that client authentication can be enabled on the server side, what do you think to add a unit test around that ?

dachaac replied:
It looks like MqttClient just got SSL support too so will see how that could be utilised for testing. There seems to be some SSL test code in vert.x core too... Let's see what I can bake.

dachaac commented:
Added some tests for SSL Client Certificate for MQTT server. Made overhaul for SSL certificates.

dachaac commented:
I wanted to make test with REQUIRE mode for client certificates and it worked fine for both two tests here. But then I wanted to try out version that client doesn't actually send certificates this caused vert.x core/netty into very weird state and it times out (or blocks whole tests to complete if debugging) and gives nice looking stack traces. Perhaps @ppatierno or @vietj could have an idea what is happening here?

It kinda feels that this special case is not handled in vert.x core?

I have pushed this testing code to https://github.com/dachaac/vertx-mqtt/tree/ssl-client-certificate-support-wip

And related parts from logs:
```
Aug 13, 2017 10:38:17 AM io.vertx.mqtt.test.server.MqttServerBaseTest
INFO: MQTT server listening on port 8883
Aug 13, 2017 10:38:17 AM io.vertx.core.net.impl.NetServerImpl
SEVERE: Client from origin /127.0.0.1:49459 failed to connect over ssl: javax.net.ssl.SSLHandshakeException: null cert chain
Aug 13, 2017 10:38:17 AM io.netty.channel.DefaultChannelPipeline onUnhandledInboundException
WARNING: An exceptionCaught() event was fired, and it reached at the tail of the pipeline. It usually means the last handler in the pipeline did not handle the exception.
io.netty.handler.codec.DecoderException: javax.net.ssl.SSLHandshakeException: null cert chain
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:442)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:248)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:129)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:642)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:565)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:479)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:441)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at java.lang.Thread.run(Thread.java:745)
Caused by: javax.net.ssl.SSLHandshakeException: null cert chain
	at sun.security.ssl.Handshaker.checkThrown(Handshaker.java:1421)
	at sun.security.ssl.SSLEngineImpl.checkTaskThrown(SSLEngineImpl.java:535)
	at sun.security.ssl.SSLEngineImpl.readNetRecord(SSLEngineImpl.java:813)
	at sun.security.ssl.SSLEngineImpl.unwrap(SSLEngineImpl.java:781)
	at javax.net.ssl.SSLEngine.unwrap(SSLEngine.java:624)
	at io.netty.handler.ssl.SslHandler$SslEngineType$2.unwrap(SslHandler.java:223)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1117)
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1039)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:411)
	... 15 more
Caused by: javax.net.ssl.SSLHandshakeException: null cert chain
	at sun.security.ssl.Alerts.getSSLException(Alerts.java:192)
	at sun.security.ssl.SSLEngineImpl.fatal(SSLEngineImpl.java:1666)
	at sun.security.ssl.Handshaker.fatalSE(Handshaker.java:304)
	at sun.security.ssl.Handshaker.fatalSE(Handshaker.java:292)
	at sun.security.ssl.ServerHandshaker.clientCertificate(ServerHandshaker.java:1804)
	at sun.security.ssl.ServerHandshaker.processMessage(ServerHandshaker.java:222)
	at sun.security.ssl.Handshaker.processLoop(Handshaker.java:969)
	at sun.security.ssl.Handshaker$1.run(Handshaker.java:909)
	at sun.security.ssl.Handshaker$1.run(Handshaker.java:906)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.security.ssl.Handshaker$DelegatedTask.run(Handshaker.java:1359)
	at io.netty.handler.ssl.SslHandler.runDelegatedTasks(SslHandler.java:1256)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1169)
	... 17 more

Aug 13, 2017 10:38:17 AM io.vertx.core.net.impl.NetServerImpl
SEVERE: Client from origin /127.0.0.1:49460 failed to connect over ssl: javax.net.ssl.SSLHandshakeException: null cert chain
Aug 13, 2017 10:38:17 AM io.netty.channel.DefaultChannelPipeline onUnhandledInboundException
WARNING: An exceptionCaught() event was fired, and it reached at the tail of the pipeline. It usually means the last handler in the pipeline did not handle the exception.
io.netty.handler.codec.DecoderException: javax.net.ssl.SSLHandshakeException: null cert chain
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:442)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:248)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:129)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:642)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:565)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:479)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:441)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at java.lang.Thread.run(Thread.java:745)
Caused by: javax.net.ssl.SSLHandshakeException: null cert chain
	at sun.security.ssl.Handshaker.checkThrown(Handshaker.java:1421)
	at sun.security.ssl.SSLEngineImpl.checkTaskThrown(SSLEngineImpl.java:535)
	at sun.security.ssl.SSLEngineImpl.readNetRecord(SSLEngineImpl.java:813)
	at sun.security.ssl.SSLEngineImpl.unwrap(SSLEngineImpl.java:781)
	at javax.net.ssl.SSLEngine.unwrap(SSLEngine.java:624)
	at io.netty.handler.ssl.SslHandler$SslEngineType$2.unwrap(SslHandler.java:223)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1117)
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1039)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:411)
	... 15 more
Caused by: javax.net.ssl.SSLHandshakeException: null cert chain
	at sun.security.ssl.Alerts.getSSLException(Alerts.java:192)
	at sun.security.ssl.SSLEngineImpl.fatal(SSLEngineImpl.java:1666)
	at sun.security.ssl.Handshaker.fatalSE(Handshaker.java:304)
	at sun.security.ssl.Handshaker.fatalSE(Handshaker.java:292)
	at sun.security.ssl.ServerHandshaker.clientCertificate(ServerHandshaker.java:1804)
	at sun.security.ssl.ServerHandshaker.processMessage(ServerHandshaker.java:222)
	at sun.security.ssl.Handshaker.processLoop(Handshaker.java:969)
	at sun.security.ssl.Handshaker$1.run(Handshaker.java:909)
	at sun.security.ssl.Handshaker$1.run(Handshaker.java:906)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.security.ssl.Handshaker$DelegatedTask.run(Handshaker.java:1359)
	at io.netty.handler.ssl.SslHandler.runDelegatedTasks(SslHandler.java:1256)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1169)
	... 17 more

Aug 13, 2017 10:38:17 AM io.vertx.mqtt.test.server.MqttServerBaseTest
INFO: MQTT server listening on port 8883
Aug 13, 2017 10:38:17 AM io.vertx.core.net.impl.NetServerImpl
SEVERE: Client from origin /127.0.0.1:49461 failed to connect over ssl: javax.net.ssl.SSLHandshakeException: null cert chain
Aug 13, 2017 10:38:17 AM io.netty.channel.DefaultChannelPipeline onUnhandledInboundException
WARNING: An exceptionCaught() event was fired, and it reached at the tail of the pipeline. It usually means the last handler in the pipeline did not handle the exception.
io.netty.handler.codec.DecoderException: javax.net.ssl.SSLHandshakeException: null cert chain
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:442)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:248)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:129)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:642)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:565)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:479)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:441)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at java.lang.Thread.run(Thread.java:745)
Caused by: javax.net.ssl.SSLHandshakeException: null cert chain
	at sun.security.ssl.Handshaker.checkThrown(Handshaker.java:1421)
	at sun.security.ssl.SSLEngineImpl.checkTaskThrown(SSLEngineImpl.java:535)
	at sun.security.ssl.SSLEngineImpl.readNetRecord(SSLEngineImpl.java:813)
	at sun.security.ssl.SSLEngineImpl.unwrap(SSLEngineImpl.java:781)
	at javax.net.ssl.SSLEngine.unwrap(SSLEngine.java:624)
	at io.netty.handler.ssl.SslHandler$SslEngineType$2.unwrap(SslHandler.java:223)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1117)
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1039)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:411)
	... 15 more
Caused by: javax.net.ssl.SSLHandshakeException: null cert chain
	at sun.security.ssl.Alerts.getSSLException(Alerts.java:192)
	at sun.security.ssl.SSLEngineImpl.fatal(SSLEngineImpl.java:1666)
	at sun.security.ssl.Handshaker.fatalSE(Handshaker.java:304)
	at sun.security.ssl.Handshaker.fatalSE(Handshaker.java:292)
	at sun.security.ssl.ServerHandshaker.clientCertificate(ServerHandshaker.java:1804)
	at sun.security.ssl.ServerHandshaker.processMessage(ServerHandshaker.java:222)
	at sun.security.ssl.Handshaker.processLoop(Handshaker.java:969)
	at sun.security.ssl.Handshaker$1.run(Handshaker.java:909)
	at sun.security.ssl.Handshaker$1.run(Handshaker.java:906)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.security.ssl.Handshaker$DelegatedTask.run(Handshaker.java:1359)
	at io.netty.handler.ssl.SslHandler.runDelegatedTasks(SslHandler.java:1256)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1169)
	... 17 more

Aug 13, 2017 10:38:17 AM io.vertx.core.net.impl.NetServerImpl
SEVERE: Client from origin /127.0.0.1:49462 failed to connect over ssl: javax.net.ssl.SSLHandshakeException: null cert chain
Aug 13, 2017 10:38:17 AM io.netty.channel.DefaultChannelPipeline onUnhandledInboundException
WARNING: An exceptionCaught() event was fired, and it reached at the tail of the pipeline. It usually means the last handler in the pipeline did not handle the exception.
io.netty.handler.codec.DecoderException: javax.net.ssl.SSLHandshakeException: null cert chain
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:442)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:248)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:129)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:642)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:565)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:479)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:441)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at java.lang.Thread.run(Thread.java:745)
Caused by: javax.net.ssl.SSLHandshakeException: null cert chain
	at sun.security.ssl.Handshaker.checkThrown(Handshaker.java:1421)
	at sun.security.ssl.SSLEngineImpl.checkTaskThrown(SSLEngineImpl.java:535)
	at sun.security.ssl.SSLEngineImpl.readNetRecord(SSLEngineImpl.java:813)
	at sun.security.ssl.SSLEngineImpl.unwrap(SSLEngineImpl.java:781)
	at javax.net.ssl.SSLEngine.unwrap(SSLEngine.java:624)
	at io.netty.handler.ssl.SslHandler$SslEngineType$2.unwrap(SslHandler.java:223)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1117)
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1039)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:411)
	... 15 more
Caused by: javax.net.ssl.SSLHandshakeException: null cert chain
	at sun.security.ssl.Alerts.getSSLException(Alerts.java:192)
	at sun.security.ssl.SSLEngineImpl.fatal(SSLEngineImpl.java:1666)
	at sun.security.ssl.Handshaker.fatalSE(Handshaker.java:304)
	at sun.security.ssl.Handshaker.fatalSE(Handshaker.java:292)
	at sun.security.ssl.ServerHandshaker.clientCertificate(ServerHandshaker.java:1804)
	at sun.security.ssl.ServerHandshaker.processMessage(ServerHandshaker.java:222)
	at sun.security.ssl.Handshaker.processLoop(Handshaker.java:969)
	at sun.security.ssl.Handshaker$1.run(Handshaker.java:909)
	at sun.security.ssl.Handshaker$1.run(Handshaker.java:906)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.security.ssl.Handshaker$DelegatedTask.run(Handshaker.java:1359)
	at io.netty.handler.ssl.SslHandler.runDelegatedTasks(SslHandler.java:1256)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1169)
	... 17 more

Tests run: 3, Failures: 1, Errors: 1, Skipped: 0, Time elapsed: 120.221 sec <<< FAILURE! - in io.vertx.mqtt.test.server.MqttServerClientCertSslTest
untrustedClient(io.vertx.mqtt.test.server.MqttServerClientCertSslTest)  Time elapsed: 0.051 sec  <<< FAILURE!
java.lang.AssertionError: Expected true
	at io.vertx.ext.unit.impl.TestContextImpl.reportAssertionError(TestContextImpl.java:432)
	at io.vertx.ext.unit.impl.TestContextImpl.assertTrue(TestContextImpl.java:282)
	at io.vertx.ext.unit.impl.TestContextImpl.assertTrue(TestContextImpl.java:288)
	at io.vertx.mqtt.test.server.MqttServerClientCertSslTest.untrustedClient(MqttServerClientCertSslTest.java:162)

clientWithoutCertificates(io.vertx.mqtt.test.server.MqttServerClientCertSslTest)  Time elapsed: 120.01 sec  <<< ERROR!
java.util.concurrent.TimeoutException: null
	at io.vertx.ext.unit.impl.TestContextImpl$Step.lambda$run$0(TestContextImpl.java:112)
	at io.vertx.ext.unit.impl.TestContextImpl$Step$$Lambda$3/1394336709.run(Unknown Source)
	at java.lang.Thread.run(Thread.java:745)
```

sauthieg commented on src/test/java/io/vertx/mqtt/test/server/MqttServerClientCertSslTest.java:
```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright 2016 Red Hat Inc.
```
I'm not sure if that matters, but shouldn't that be set to `2017` ?

dachaac replied:
Now it was just a copy -- should it have something like:
"Copyright © 2017 The Eclipse Foundation. All Rights Reserved."

@ppatierno can probably comment on this one?

ppatierno replied:
Having Red Hat is ok. For now let's leave 2016, then I'll change it. No worries ;)

dachaac commented:
Note: for REQUIRE mode handling case without client sending certificate I have created issue for vert.x:
https://github.com/eclipse/vert.x/issues/2104

Require also works but didn't figure out how to write test case to capture its handling in vertx-mqtt server side as I didn't spot way to capture the exception.

Apart from that I believe this could go in and if fix comes for vert.x/netty then additional test case for this particular special case can be added.

@ppatierno please consider PR for review and for possible merge.

dachaac commented:
@ppatierno they were generated by openssl's CA functionality automatically. Removed them and added matching .gitignore's so they are not added in future.

ppatierno commented on src/test/resources/tls/client-trusted-root-ca/.gitignore:
```diff
@@ -0,0 +1 @@
+*.old
```
It's better to use the main project .gitignore file with something like : 

**/*.old

instead of more .gitignore files here.

ppatierno commented on src/test/resources/tls/client-untrusted-root-ca/.gitignore:
```diff
@@ -0,0 +1 @@
+*.old
```
Ditto as above

ppatierno commented on src/test/resources/tls/server-root-ca/.gitignore:
```diff
@@ -0,0 +1 @@
+*.old
```
Ditto as above

dachaac commented:
@ppatierno moved .gitignore's rules to root folder's .gitignore as suggested.

ppatierno commented on src/main/java/io/vertx/mqtt/MqttEndpoint.java:
```diff
@@ -41,6 +44,31 @@
+   * @return the remote address for this socket
+   */
+  @CacheReturn
+  SocketAddress remoteAddress();
```
Exposing `remoteAddress` is not related to the purpose of this PR on SSL right ? Why do you need this ?

dachaac replied:
It is useful for logging purposes. I can create another PR if you wish for that functionality.

ppatierno commented on src/main/java/io/vertx/mqtt/MqttEndpoint.java:
```diff
@@ -41,6 +44,31 @@
+   * @return the local address for this socket
+   */
+  @CacheReturn
+  SocketAddress localAddress();
```
Ditto as above.

dachaac replied:
It is useful for logging purposes. I can create another PR if you wish for that functionality.

ppatierno replied:
@dachaac yes. It's better having a different PR for that. Thanks !

dachaac replied:
@ppatierno Created new PR for that: https://github.com/vert-x3/vertx-mqtt/pull/56

Will remove it from this PR now.

dachaac commented:
@ppatierno This PR updated with local/remoteAddress() being moved to own PR.

ppatierno commented:
@dachaac I have merged the PR related to the local/remoteAddress(). Now this PR has some conflicts due to that merge. You need to updated your fork in order to fix them having the new MqttEndpoint

dachaac commented:
@ppatierno Rebased this PR on top of master.

dachaac commented:
@ppatierno Is there something needing more work or would this be ready for merge?

vietj commented:
btw this PR for core might interest you for testing purposes https://github.com/eclipse/vert.x/pull/2129

dachaac commented:
@vietj Is this something that you were thinking of: https://github.com/dachaac/vertx-mqtt/commits/ssl-client-with-improved-tests

Mainly commits: https://github.com/dachaac/vertx-mqtt/commit/ac534e5c774fa7e7153f2f94c4af2a5511012e84 and https://github.com/dachaac/vertx-mqtt/commit/8185a98cd46e6df6f7db02cd3e833edcbf916428.

Note: Created separate PR https://github.com/vert-x3/vertx-mqtt/pull/62 for waiting that MQTT server is up'n'running before actually starting tests. That commit is in that branch for testing purposes.

vietj commented:
can you provide a diff ?

> On Sep 24, 2017, at 12:12 PM, Vesa Jääskeläinen <notifications@github.com> wrote:
> 
> @vietj <https://github.com/vietj> Is this something that you were thinking of: https://github.com/dachaac/vertx-mqtt/commits/ssl-client-with-improved-tests <https://github.com/dachaac/vertx-mqtt/commits/ssl-client-with-improved-tests>
> Mainly commits: dachaac/vertx-mqtt@ac534e5 <https://github.com/dachaac/vertx-mqtt/commit/ac534e5c774fa7e7153f2f94c4af2a5511012e84> and dachaac/vertx-mqtt@8185a98 <https://github.com/dachaac/vertx-mqtt/commit/8185a98cd46e6df6f7db02cd3e833edcbf916428>.
> 
> Note: Created separate PR #62 <https://github.com/vert-x3/vertx-mqtt/pull/62> for waiting that MQTT server is up'n'running before actually starting tests. That commit is in that branch for testing purposes.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/vert-x3/vertx-mqtt/pull/42#issuecomment-331700217>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AANxihEfeirsZWA0P_qJCiqcDluIMbGVks5slismgaJpZM4OoKfI>.
>

dachaac commented:
@vietj Related changes for MQTT server: https://github.com/dachaac/vertx-mqtt/commit/ac534e5c774fa7e7153f2f94c4af2a5511012e84.

And diff between request and require tests:
```
$ diff -pruN MqttServerClientCertSslTest.java MqttServerClientCertRequiredSslTest.java
--- MqttServerClientCertSslTest.java	2017-09-24 13:04:58.000000000 +0300
+++ MqttServerClientCertRequiredSslTest.java	2017-09-24 13:05:13.000000000 +0300
@@ -47,9 +47,9 @@ import java.security.cert.X509Certificat
  * MQTT server testing using SSL/TLS with client certificates
  */
 @RunWith(VertxUnitRunner.class)
-public class MqttServerClientCertSslTest extends MqttServerBaseTest {
+public class MqttServerClientCertRequiredSslTest extends MqttServerBaseTest {

-  private static final Logger log = LoggerFactory.getLogger(MqttServerClientCertSslTest.class);
+  private static final Logger log = LoggerFactory.getLogger(MqttServerClientCertRequiredSslTest.class);

   private Async async;

@@ -63,7 +63,6 @@ public class MqttServerClientCertSslTest

   @Before
   public void before(TestContext context) {
-
     PemKeyCertOptions pemKeyCertOptions = new PemKeyCertOptions()
       .setKeyPath("tls/server-key.pem")
       .setCertPath("tls/server-cert.pem");
@@ -76,7 +75,7 @@ public class MqttServerClientCertSslTest
       .setKeyCertOptions(pemKeyCertOptions)
       .setTrustOptions(pemTrustOptions)
       .setSsl(true)
-      .setClientAuth(ClientAuth.REQUEST);
+      .setClientAuth(ClientAuth.REQUIRED);

     this.setUp(context, options);
   }
@@ -150,14 +149,58 @@ public class MqttServerClientCertSslTest

       this.async.await();

-      client.disconnect();
-
-      context.assertTrue(this.clientConnectedWithSsl);
-      context.assertFalse(this.receivedClientCertificateValidated);
-      context.assertNull(receivedClientPeerCertificates);
+      context.fail("We should get exception!");

     } catch (MqttException e) {
+      if (e.getCause() instanceof SSLHandshakeException) {
+        // Note: It is impossible to diagnose exception type in portable way
+
+        context.assertFalse(this.clientConnectedWithSsl);
+        context.assertFalse(this.receivedClientCertificateValidated);
+        context.assertNull(receivedClientPeerCertificates);
+
+        this.async.complete();
+      } else {
+        context.fail(e);
+      }
+    } catch (Exception e) {
       context.fail(e);
+    }
+  }
+
+  @Test
+  public void clientWithoutCert(TestContext context) {
+
+    resetClientInfo();
+
+    this.async = context.async();
+
+    try {
+
+      MemoryPersistence persistence = new MemoryPersistence();
+      MqttClient client = new MqttClient(String.format("ssl://%s:%d", MQTT_SERVER_HOST, MQTT_SERVER_TLS_PORT), "12345", persistence);
+
+      MqttConnectOptions options = new MqttConnectOptions();
+      client.connect(options);
+
+      client.publish(MQTT_TOPIC, MQTT_MESSAGE.getBytes(), 0, false);
+
+      this.async.await();
+
+      context.fail("We should get exception!");
+
+    } catch (MqttException e) {
+      if (e.getCause() instanceof SSLHandshakeException) {
+        // Note: It is impossible to diagnose exception type in portable way
+
+        context.assertFalse(this.clientConnectedWithSsl);
+        context.assertFalse(this.receivedClientCertificateValidated);
+        context.assertNull(receivedClientPeerCertificates);
+
+        this.async.complete();
+      } else {
+        context.fail(e);
+      }
     } catch (Exception e) {
       context.fail(e);
     }
```

What I was wondering is should there be two different exception handlers registered -- now they are the same.

I will probably rename class for request test to include Request in there too like in require one.

dachaac commented:
@ppatierno Is there something you want to be done for this PR or could it be merged?

ppatierno commented on src/test/resources/tls/SSL.md:
```diff
@@ -0,0 +1,158 @@
+
+Client certificates and private keys are all stored in PEM format and in JKS format.
+
+### Trusted Root CA for SSL Client Certificates
```
Just a question, why are you using two different root CAs for signing server and client certificates ? Can we use a single root CA for simplicity ?

vietj commented:
for the sake of simplicity perhaps you can reuse the certificates / keys from `io.vertx:vertx-core` test jar ?

dachaac commented:
@vietj certificates and keys could be a copy. However they would anyway drift a bit during time so from that point of view there is no difference. Trying to keep them in sync gives more work than what is needed for vertx-mqtt testing.

What is different is that expiration date is a bit longer in these certificates so that automated tests do not need to be updated yearly due to expired certificates.

Another difference is that these use SHA256 as required by today's standards.

In vertex-core there is only one client root CA which this setup mitigates.

Also setup in vertx-core is more complex due to fact that it needs to make sure that SSL verification is done properly in multiple different cases. Thou there may be something that could be improved for certificate expiration testing if I recall properly.

In SSL.md vertx-mqtt's test cases are documented and instructions are given should there be a need to re-create them in future. I personally had a bit challenges trying to understand terminology used in vertx-core/ssl.txt describing what is used for what. Thus I created in my opinion more descriptive SSL.md. Used markdown so that it is formatted nicely in github.

vietj commented:
I mean to reuse them directly from the jar file.

Perhaps that ssl.txt is not perfect and not in markdown format (yet), however I'm concerned by the fact that  it makes two set of certificates / pair to maintain

dachaac commented:
@vietj personally I do not see that as a problem. Certificates are not that hard to set up. It has been now done and even with instructions how to re-do it if needed. 

Maintainer of the software of course can make a choice on how to develop the software. 

Personally I see it more challenging when trying to understand how these tests work if you split these to another project. If there is a change in vertx-core on these then these tests break down. I personally like self-contained approach.

If splitting is needed then vertx-core needs to be first updated with suitable certificates and keys.

vietj commented:
@dachaac why is testing with pem and p12 store necessary ? this is tested in vertx core and beside the format, these are the same key

dachaac commented:
@vietj p12 is only intermediate format to convert openssl generated private key and certificate chain to Java keytool format (jks). It is not used in testing as such.

jks is used for Eclipse Paho.

pkcs8 (pem) is used in vertx.

sophokles73 commented:
Is there any particular reason why this PR hasn't made it into vertx-mqtt yet? The Eclipse Hono project would really like to take advantage of being able to authenticate clients using X509 certificates ...

vietj commented:
one issue from my perspective is that it could reuse the same certificates from the vertx-core testsuite and keep things simpler to maintain instead of duplicating the management of test certs and keys. There is some arguing about the markdown format for describing the how-to but this could be resolved by improving vertx-core with that respect.

vietj commented:
@ppatierno also seems to request changes

dachaac commented:
@vietj & @ppatierno - please state clearly if you are discussing on matter or **requiring** a change related to this PR. I believe I have answered to all your comments.

@vietj Did you notice the replies I wrote on the certificate test material matter?

dachaac commented:
@vietj I can give it a try.

dachaac commented:
@vietj I am having problems getting vert.x core test-sources to be used as a resource only during test execution. It seems that it gets both classes and other resources inside the test execution making test execution to fail.
```
Tests run: 8, Failures: 0, Errors: 8, Skipped: 0, Time elapsed: 0.184 sec <<< FAILURE! - in io.vertx.mqtt.test.server.MqttServerSubscribeTest
subscribeFailure(io.vertx.mqtt.test.server.MqttServerSubscribeTest)  Time elapsed: 0.155 sec  <<< ERROR!
java.util.ServiceConfigurationError: io.vertx.core.spi.VerticleFactory: Provider io.vertx.test.core.ClasspathVerticleFactory not found
...
```
This is what I tried in pom:
```
    <dependency>
      <groupId>io.vertx</groupId>
      <artifactId>vertx-core</artifactId>
      <version>${stack.version}</version>
      <classifier>test-sources</classifier>
      <type>test-jar</type>
      <scope>test</scope>
    </dependency>
```
Any ideas for this?

I believe one solution could be to move all vertx-core test resources to own jar that would be re-used but not sure you were thinking about that? As per:
https://maven.apache.org/plugins/maven-jar-plugin/examples/create-test-jar.html

vietj commented:
vertx-web uses

```
    <dependency>
      <groupId>io.vertx</groupId>
      <artifactId>vertx-core</artifactId>
      <type>test-jar</type>
      <scope>test</scope>
    </dependency>
```

vietj commented:
I think you don't need dependency on test sources

dachaac commented:
@vietj Sure. It is referenced in this PR. It is rebaseable to latest version:
https://github.com/dachaac/vertx-mqtt/tree/ssl-client-certificate-support

But failures appear even with the master with that pom change I made.

Reason why I referenced to test sources jar was because those files were there.

sophokles73 commented:
@vietj @dachaac any progress on this issue?

vietj commented:
@dachaac I've rebased on latest master https://github.com/vert-x3/vertx-mqtt/tree/ssl-client-certificate-support and it works fine for me, can you have a look please ?

dachaac commented:
@vietj I'll try it out tonight

dachaac commented:
I spent couple of hours on some weekend and there was logical problem what certificate would be ok and had problems for getting the certs and keys loaded from tests jar.

vietj commented:
@dachaac thanks, sorry for the long delay

dachaac commented:
@vietj umh. It seems to be still using certs in vertx-mqtt?

vietj commented:
I thought you made changes

vietj commented:
ok so I need to try to use the cert from vertx-core test jar right ?

dachaac commented:
@vietj Just figured out why my testing wasn't working. As it is more or less impossible to run vert.x tests successfully I was skipping them. But I noticed that I used wrong skipping method. I was skipping whole tests including building them and populating them in m2 cache.

Was doing:

mvn install -Dmaven.test.skip=true

instead of:

mvn install -DskipTests

Now I get my modified CA stores also to other side. Now I just need to figure out what I want to change in vertx-core's TLS certificate material as they do not feature everything for proper test. (alternatively I scale down testing and just use what is there)

Currently it is missing client certs which are signed by an CA so that Eclipse Paho MQTT client could send them. Also CA names are a bit confusing.

vietj commented:
@dachaac I've pushed an update that unpack the vertx-core tls resources and use them, some TLS test pass and some fail (with trusted CA)

dachaac commented:
Note: Dropped Eclipse Paho upgrade as it is not related to this pr.

dachaac commented:
@vietj and @ppatierno please review.

ppatierno commented on src/test/java/io/vertx/mqtt/test/server/MqttServerBaseTest.java:
```diff
@@ -103,4 +107,42 @@ protected void endpointHandler(MqttEndpoint endpoint, TestContext context) {
+   *
+   * @param trustStoreName Trust storename in classpath format
+   * @param keyStoreName Key store name in classpath format
+   * @return
```
missing return description

vietj commented:
I will review the vertx-core PR asap, it looks like a great contribution.

vietj commented:
@dachaac the vertx-core PR has been merged and a SNAPSHOT is published

vietj commented:
can we merge this PR @ppatierno ?

vietj commented:
awesome :-)
